receiver/awscontainerinsightreceiver/internal/gpu/decorator.go

```diff
@@ -0,0 +1,204 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package gpu
+
+import (
+	"context"
+	"errors"
+
+	ci "github.com/open-telemetry/opentelemetry-collector-contrib/internal/aws/containerinsight"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/awscontainerinsightreceiver/internal/stores"
+	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/pmetric"
+	"go.uber.org/zap"
+	"golang.org/x/exp/maps"
+)
+
+const (
+	gpuUtil        = "DCGM_FI_DEV_GPU_UTIL"
+	gpuMemUtil     = "DCGM_FI_DEV_FB_USED_PERCENT"
+	gpuMemUsed     = "DCGM_FI_DEV_FB_USED"
+	gpuMemTotal    = "DCGM_FI_DEV_FB_TOTAL"
+	gpuTemperature = "DCGM_FI_DEV_GPU_TEMP"
+	gpuPowerDraw   = "DCGM_FI_DEV_POWER_USAGE"
+)
+
+var _ stores.CIMetric = (*gpuMetric)(nil)
+
+var metricToUnit = map[string]string{
+	gpuUtil:        "Percent",
+	gpuMemUtil:     "Percent",
+	gpuMemUsed:     "Bytes",
+	gpuMemTotal:    "Bytes",
+	gpuTemperature: "None",
+	gpuPowerDraw:   "None",
+}
+
+type gpuMetric struct {
+	// key/value pairs that are typed and contain the metric (numerical) data
+	fields map[string]any
+	// key/value string pairs that are used to identify the metrics
+	tags map[string]string
+}
+
+func newResourceMetric(mType string, logger *zap.Logger) *gpuMetric {
+	metric := &gpuMetric{
+		fields: make(map[string]any),
+		tags:   make(map[string]string),
+	}
+	metric.tags[ci.MetricType] = mType
+	return metric
+}
+
+func (gr *gpuMetric) GetTags() map[string]string {
+	return gr.tags
+}
+
+func (gr *gpuMetric) GetFields() map[string]any {
+	return gr.fields
+}
+
+func (gr *gpuMetric) GetMetricType() string {
+	return gr.tags[ci.MetricType]
+}
+
+func (gr *gpuMetric) AddTags(tags map[string]string) {
+	for k, v := range tags {
+		gr.tags[k] = v
+	}
+}
+
+func (gr *gpuMetric) HasField(key string) bool {
+	return gr.fields[key] != nil
+}
+
+func (gr *gpuMetric) AddField(key string, val any) {
+	gr.fields[key] = val
+}
+
+func (gr *gpuMetric) GetField(key string) any {
+	return gr.fields[key]
+}
+
+func (gr *gpuMetric) HasTag(key string) bool {
+	return gr.tags[key] != ""
+}
+
+func (gr *gpuMetric) AddTag(key, val string) {
+	gr.tags[key] = val
+}
+
+func (gr *gpuMetric) GetTag(key string) string {
+	return gr.tags[key]
+}
+
+func (gr *gpuMetric) RemoveTag(key string) {
+	delete(gr.tags, key)
+}
+
+// GPU decorator acts as an interceptor of metrics before the scraper sends them to the next designated consumer
+type decorateConsumer struct {
+	containerOrchestrator string
+	nextConsumer          consumer.Metrics
+	k8sDecorator          Decorator
+	logger                *zap.Logger
+}
+
+func (dc *decorateConsumer) Capabilities() consumer.Capabilities {
+	return consumer.Capabilities{
+		MutatesData: true,
+	}
+}
+
+func (dc *decorateConsumer) ConsumeMetrics(ctx context.Context, md pmetric.Metrics) error {
+	resourceTags := make(map[string]string)
+	rms := md.ResourceMetrics()
+	for i := 0; i < rms.Len(); i++ {
```

Comment:
Probably better to move this entire loop into the `ConvertToFieldsAndTags()` method.  It would be the inverse of the existing `ConvertToOTLPMetrics()` method.  Take in `pmetrics.Metrics` and return an array of fields/tags pairs (or wrap in  struct).

Then here in decoratorConsumer, iterate through again and call the decorator.

Reply:
Oh, interesting point, we are only using `K8sDecorator` for tags in this case, we don't need the fields at all.  Hmm... seems like we could have done some additional refactoring then... separate `K8sDecorator` into a series of sub-decorators - cadvisor would use all but dcgm would only use the couple that it needs.  I won't push on it now though.

Also worth noting that `Cadvisor`/`K8sDecorator` only work this way because they were ported from the original CWAgent.  If we were writing from scratch, we would probably operate on OTLP metrics directly.

Reply:
Ah that makes sense. I was wondering why existing decorator wouldn't just start with OTLP metrics. 

Breaking k8sdecorator will be an interesting one. It currently depends on 2 stores (`podstore` and `servicestore`) that it uses to decorate. `podstore` is the main one with most of data (such as node and container) available at the pod level. Making these pluggable for different needs will be nice and reduce unnecessary overhead and processing.